src/webgpu/api/operation/memory_sync/texture/same_subresource.spec.ts

```diff
@@ -2,21 +2,465 @@ export const description = `
 Memory Synchronization Tests for Texture: read before write, read after write, and write after write to the same subresource.
 
 - TODO: Test synchronization between multiple queues.
+- TODO: Test depth/stencil attachments.
 `;
 
+import { SkipTestCase } from '../../../../../common/framework/fixture.js';
 import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import { assert, memcpy, unreachable } from '../../../../../common/util/util.js';
+import { EncodableTextureFormat } from '../../../../capability_info.js';
 import { GPUTest } from '../../../../gpu_test.js';
+import { align } from '../../../../util/math.js';
+import { getTextureCopyLayout } from '../../../../util/texture/layout.js';
+import {
+  kTexelRepresentationInfo,
+  PerTexelComponent,
+} from '../../../../util/texture/texel_data.js';
 
 import {
   kOperationBoundaries,
   kBoundaryInfo,
   kAllReadOps,
   kAllWriteOps,
   checkOpsValidForContext,
+  Op,
+  OperationBoundary,
+  OperationContext,
+  kOpInfo,
+  kOperationContexts,
 } from './texture_sync_test.js';
 
 export const g = makeTestGroup(GPUTest);
 
+class TextureSyncTestHelper {
+  // We start at the queue context which is top-level.
+  private currentContext: OperationContext = 'queue';
+
+  // Set based on the current context.
+  private queue: GPUQueue;
+  private commandEncoder?: GPUCommandEncoder;
+  private computePassEncoder?: GPUComputePassEncoder;
+  private renderPassEncoder?: GPURenderPassEncoder;
+  private renderBundleEncoder?: GPURenderBundleEncoder;
+
+  private device: GPUDevice;
+  private texture: GPUTexture;
+
+  private encodedCommands: GPURenderBundle[] | GPUCommandBuffer[] = [];
```

Comment:
does this ever contain more than 1 item? (related comment below about not flushing this every time)